### PR TITLE
Add docs to `ReadStorage` and `WriteStorage`

### PR DESCRIPTION
**Pull request by torkleyy**

### CODE REVIEW

File: src/storage/data.rs [hunk 1]
```diff
@@ -4,6 +4,96 @@ use {Component, EntitiesRes, Storage};
 use storage::MaskedStorage;
 
 /// A storage with read access.
+///
+/// This is just a typedef for a fetched component storage.
```
Comment (WaDelma) [inline src/storage/data.rs]: type alias

File: src/storage/data.rs [hunk 2]
```diff
@@ -4,6 +4,96 @@ use {Component, EntitiesRes, Storage};
...
+///     fn run(&mut self, (ent, pos): Self::SystemData) {
+///         for (ent, pos) in (&*ent, &pos).join() {
+///             println!("Entitiy with id {} has a position of {:?}", ent.id(), pos);
+///         }
+///     }
+/// }
+/// ```
+///
+/// These operations can't mutate anything; if you want to do
+/// insertions or modify components, you need to use `WriteStorage`.
```
Comment (WaDelma) [inline src/storage/data.rs]: or use `LazyUpdate`

File: src/storage/data.rs [hunk 3]
```diff
@@ -27,6 +117,54 @@ where
 }
 
 /// A storage with read and write access.
+///
+/// In addition to `ReadStorage`, a storage with mutable access allows to:
```
Comment (WaDelma) [inline src/storage/data.rs]: Additionally to what `ReadStorage` can do a storage with mutable access allows:

File: src/storage/data.rs [hunk 4]
```diff
@@ -27,6 +117,54 @@ where
...
+/// let entity = world.create_entity()
+///     .with(Pos(0.1))
+///     .build();
+/// # let mut pos_storage = world.write::<Pos>();
+///
+/// let res = pos_storage.insert(entity, Pos(4.0));
+/// assert_eq!(res, InsertResult::Updated(Pos(0.1)));
+/// ```
+///
+/// There's also an Entry-API similar to the one provided by `std`.
```
Comment (WaDelma) [inline src/storage/data.rs]: provided by `std::collections::HashMap`.

**Comment by LinearZoetrope** [comment]
LGTM, though there probably should be expanded ones for `Entities` (mostly detailing the needed reborrow which is peculiar compared to `Read/WriteStorage` as well as some note about using the flavors of `Fetch` to grab Resources in `SystemData`.

File: src/storage/data.rs (hunk 3 above)
Comment (torkleyy) [reply src/storage/data.rs]: My sentence doesn't make any sense at all, thanks for the suggestion.

**Comment by LinearZoetrope** [comment]
It may be worth renaming it to RwStorage, but obviously that would be breaking.

**Comment by Xaeroxe** [comment]
Usually what I see happen is people will write code that breaks the borrow checker rules, and rather than fix that they instead add a ReadStorage to the system data and use that instead.  Makes the borrow checker shut up, but it instead panics at run time now.

File: src/storage/data.rs [hunk 5]
```diff
@@ -4,6 +4,99 @@ use {Component, EntitiesRes, Storage};
...
+/// #
+/// # let mut world = World::new(); world.register::<Pos>(); world.register::<Vel>();
+/// # let pos_storage = world.read::<Pos>();
+/// # let vel_storage = world.read::<Vel>();
+/// (&pos_storage, &vel_storage).join()
+/// # ;
+/// ```
+///
+/// This joins the position and the velocity storage, thus only components
+/// that exist in both storages are iterated over.
```
Comment (pengowen123) [inline src/storage/data.rs]: I think  the wording here should be changed. `join` iterates over the components of entities that have both components.

**Comment by torkleyy** [comment]
Thanks for the suggestion, please review again.

**Comment by pengowen123** [comment]
Looks good 👍.

**Comment by torkleyy** [comment]
Thanks for all the reviews!

bors r+

**Pull request merged**